### PR TITLE
MRG: in core, enable downsample within select

### PR DESCRIPTION
**Pull request by bluegenes**
Attempting #1292 in order to move forward https://github.com/sourmash-bio/sourmash_plugin_branchwater/pull/134

Modifies `Signature` `Select` to downsample automatically.

- for scaled sketches, while checking ksize, we also retain only sketches that have the right scaled or can be downsampled (scaled <= selection.scaled())
- next, we iterate through the sketches and downsample any where scaled < selection.scaled()

I thought the change was needed in `Signature` `Select` rather than `SigStore` `Select`, but lmk if I missed something!

### CODE REVIEW

**Comment by bluegenes** [comment]
I've currently modified `Signature` `Select` to downsample automatically.
- first, for scaled sketches, we retain only sketches that can be downsampled (`scaled <= selection.scaled()`)
- next, we iterate through the sketches and downsample any where `scaled < selection.scaled()`

I went back and forth about whether @luizirber was suggesting to put the downsample logic over in `SigStore` `Select` instead -- it does work there, but I figured we only want to downsample when actually loading the signature, which would likely be done via `Signature` `Select`, right?

e.g. for branchwater, we need automatic downsampling from a collection to build an index. Looks like when we go to load signatures for index building, we use `sig_for_dataset` in `Collection`, which has the following relevant lines:

```
 let selection = Selection::from_record(&self.manifest[dataset_id as usize])?;
 let sig = self.storage.load_sig(match_path)?.select(&selection)?;
```

I assume the final `select` there is using `Signature` `Select`, which is why I put the logic there. 

@luizirber is this the right approach? What would you suggest I do differently? 

I could also add some `scaled` logic to the `Manifest` Select to keep only compatible scaled:

```
            // keep compatible scaled
            valid = if let Some(scaled) = selection.scaled() {
                 if let Some(row_scaled) = row.scaled() {
                     valid && row_scaled <= scaled
                 } else {
                     valid
                 }
            };

... would this be worthwhile?

cc @ctb

**Comment by bluegenes** [comment]
@luizirber @ctb ready for review

**Review by ctb — APPROVED** [review]
This looks good to me although I'd suggest waiting for Luiz before merging :)

File: src/core/src/signature.rs [hunk 1]
```diff
@@ -979,4 +1001,47 @@ mod test {
...
+
+        // create Selection object
+        let mut selection = Selection::default();
+        selection.set_scaled(2000);
+        // iterate and check scaled
+        for sig in &sigs {
+            let modified_sig = sig.clone().select(&selection).unwrap();
+            for sketch in modified_sig.sketches() {
+                if let Sketch::MinHash(mh) = sketch {
+                    eprintln!("scaled: {:?}", mh.scaled());
```
Comment (luizirber) [inline src/core/src/signature.rs]:
```suggestion
                    dbg!("scaled: {:?}", mh.scaled());
```

**Review by luizirber — APPROVED** [review]
Nice!

Considering the `Select::select` function signature:
```
fn select(mut self, selection: &Selection) -> Result<Self, Error>
```
I think it is OK to downsample it at the `Signature` level, and not `SigStore`

**Comment by bluegenes** [comment]
Note that for `sourmash_plugin_branchwater` compatibility, we need:

- `byteorder` = "1.4.3"
- `wasm-bindgen` = "0.2.89"
- `once_cell` = "1.18.0"

**Pull request merged**